### DISCUSSION TITLE
VS Code Host doesn't deliver Rename event to all clients instantiated for each workspace folder in a multi-root workspace

### DISCUSSION THREAD

**Discussion by selim-nasir** [microsoft/vscode-discussions#1002] [Extension Development QnA] [answered]
Dear community,

I've ended up here as suggested by VS Code team regarding this issue: https://github.com/microsoft/vscode/issues/200372

Would you take your time to look at the issue and comment on regarding possible root-cause of the issue?

Thanks

**Comment by isidorn** [comment]
Can you please share reproducible steps? Do you have a simple extension we can run and see this behaviour?

**Comment by selim-nasir** [comment]
Hi @isidorn thanks a lot for looking into the issues. U'tely our extension as well as language server implementation is not publicly available. But maybe providing following scenario may help here:

**Prerequisites**
-	Custom Language Server implementation that can handle renaming of elements in any language, in our case it’s Yaml.  
-	Custom VS Code extension, that creates vscode-languageclient.LanguageClient.LanguageClient instances per workspace folder in a VS Code workspace. Each client creates an instance of custom Language Server.
- 	Extension is installed in VS Code.

**Repro Steps**
-	Open a multi-rooted VS Code workspace that contains code files in different workspace folders.
-	Create a code file in one of workspace folders with some arbitrary content and reference the file from all other workspace folders (foo.yml inside folder A in the example below). In our case, we have single config file per workspace folder that contains list of of files included in the workspace folder. In this config file, files from other workspace folders can be referenced. This config file and files section inside are specific to our project.
-	In other folders B and C, use elements from the common file.
-	Rename an element in the common file which is used in other folders.

Expected Behavior
                All rename changes in the common file are reflected in the files, that use renamed elements in other workspace folders. 
                Language Server instances that are responsible for workspace folders B, C should be notified of renaming in the foo.yml (in A) through LanguageClients. 
                
Current Behavior
                Only one of folders (either B or C) arbitrarily gets notification. VS Code host notifies only one of LanguageClients. (This is not the case with hover event; it works as expected)


Example VS Code workspace structure
                A
                               a.config.yml
                                               files:
                                                               - foo.yml
                                                                               <<defines>> elementX ----<<rename as>>--> elemW
                                                               - bar.yml
                B
                               b.config.yml
                                               files:
                                                              - baz.yml
                                                                               <<uses>> elementX (LanguageClient for B as well as Language Server are not notified. The name elementX becomes stale)
                                                               - ../A/foo.yml
                C
                               c.config.yml
                                                               - xyz.yml
                                                                               <<uses>> elementX ---<<renamed as>>---> elemW
                                                               - ../A/foo.yml

**Reply by isidorn** [reply]
@dbaeumer can you please help here? Or point @selim-nasir in the right direction. Thanks!

**Comment by dbaeumer** [comment, marked as answer]
Renames events are not propagated to more than one provider. This is how things work in VS Code and the behavior is independent from LSP. In VS Code language providers that compute results that are easily combinable (e.g. hover, find all references, go to ...) will all be notified. Providers were results are not combinable (for example rename) usually the first provider producing a result determines the result of the whole operation. The corresponding API documentation says for rename

```
		 * Multiple providers can be registered for a language. In that case providers are sorted
		 * by their {@link languages.match score} and asked in sequence. The first provider producing a result
		 * defines the result of the whole operation.
```

See https://github.com/microsoft/vscode/blob/8839e62ada712a958e76e9c5cd62a41176c51501/src/vscode-dts/vscode.d.ts#L14053

If you want to propagate renames to different language servers you need to implement that yourself. Especially since this need some dependency analysis between the project which VS Code can not know about. 

In your case I would do this by implementing a special middle ware in the language client (see https://github.com/microsoft/vscode-languageserver-node/blob/2d33852e06ff9f180abd528be464f02b9bbd3079/client/src/common/rename.ts#L26 and https://github.com/microsoft/vscode-languageserver-node/blob/2d33852e06ff9f180abd528be464f02b9bbd3079/client/src/common/client.ts#L356)

**Reply by selim-nasir** [reply]
Thanks a lot @dbaeumer for your detailed answer and implementation suggestion.
